File: leo2-node/src/main/java/org/deku/leo2/node/data/sync/EntitySync.java
package org.deku.leo2.node.data.sync;

import org.deku.leo2.messaging.activemq.ActiveMQContext;
import org.deku.leo2.node.data.entities.master.*;
import sx.Disposable;
import sx.LazyInstance;

import javax.persistence.EntityManagerFactory;

/**
 * Supervising sync class
 * Created by masc on 19.06.15.
 */
public class EntitySync implements Disposable {
    private static final LazyInstance<EntitySync> mInstance = new LazyInstance(EntitySync::new);

    public static EntitySync instance() { return mInstance.get(); };

    private EntityConsumer mEntityConsumer;
    private EntityManagerFactory mEntityManagerFactory;

    public void setEntityManagerFactory(EntityManagerFactory entityManagerFactory) {
        mEntityManagerFactory = entityManagerFactory;
    }

    /** c'tor */
    private EntitySync() { }

    public void start() {
        mEntityConsumer = new EntityConsumer(ActiveMQContext.instance(), mEntityManagerFactory);
        mEntityConsumer.request(Station.class);
        mEntityConsumer.request(Country.class);
        mEntityConsumer.request(HolidayCtrl.class);
        mEntityConsumer.request(Route.class);
        mEntityConsumer.request(Sector.class);
    }

    public void stop() {
        mEntityConsumer.dispose();
    }

    @Override
    public void dispose() {
        this.stop();
    }
}

File: leoz-central/src/main/java/org/deku/leoz/central/data/sync/EntitySync.kt
package org.deku.leoz.central.data.sync

import org.apache.commons.logging.Log
import org.apache.commons.logging.LogFactory
import org.deku.leoz.messaging.activemq.ActiveMQContext
import org.deku.leoz.node.data.sync.EntityPublisher
import sx.Disposable
import sx.LazyInstance
import sx.event.EventDelegate
import java.sql.Timestamp
import javax.inject.Named

import javax.persistence.EntityManagerFactory
import kotlin.properties.Delegates

/**
 * Supervising sync class
 * TODO: move to EntitySyncConfiguration
 * Created by masc on 19.06.15.
 */
@Named
class EntitySync private constructor() : Disposable {
    //region Singleton
    companion object Singleton {
        private val instance = EntitySync()
        @JvmStatic fun it(): EntitySync {
            return instance;
        }
    }
    //endregion

    /** Log */
    private val log = LogFactory.getLog(this.javaClass)
    /** Entity publisher */
    private var entityPublisher: EntityPublisher by Delegates.notNull()

    /** Injected: jpa entity manager factory  */
    var entityManagerFactory: EntityManagerFactory by Delegates.notNull()
    /** Injected: database sync instance */
    var databaseSync: DatabaseSync by Delegates.notNull()

    /** Database sync event listener */
    private val databaseSyncEvent = object : DatabaseSync.EventListener {
        override fun onUpdate(entityType: Class<out Any?>, currentTimestamp: Timestamp?) {
            // Publish notification to consumers on databasesync update
            entityPublisher.publish(entityType, currentTimestamp);
        }
    }

    /** Start entity sync */
    fun start() {
        // Register event
        this.databaseSync.eventDelegate.add(databaseSyncEvent)

        // Configure and start publisher
        this.entityPublisher = EntityPublisher(
                ActiveMQContext.instance,
                this.entityManagerFactory)

        this.entityPublisher.start()
    }

    /** Dispose */
    override fun dispose() {
        this.entityPublisher.stop()
    }
}
